File: frontend/src/types/patients14.ts
/** Domain types for patients / PatientRecord14 */
export type PatientRecord14Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface PatientRecord14 {
  id: string;
  externalCode: string;
  displayName: string;
  status: PatientRecord14Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface PatientRecord14CreateRequest {
  externalCode: string;
  displayName: string;
  status: PatientRecord14Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface PatientRecord14ListResponse {
  items: PatientRecord14[];
  total?: number;
}

export interface PatientRecord14StatsResponse {
  activeCount: number;
}

export function isPatientRecord14Active(rec: PatientRecord14): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatPatientRecord14Label(rec: PatientRecord14): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function comparePatientRecord14ByPriority(a: PatientRecord14, b: PatientRecord14): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}

File: backend/internal/service/patients_14.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// PatientRecord14Service encapsulates business rules for PatientRecord14.
type PatientRecord14Service struct {
	repo *repository.PatientRecord14Repository
}

// NewPatientRecord14Service wires the service.
func NewPatientRecord14Service(repo *repository.PatientRecord14Repository) *PatientRecord14Service {
	return &PatientRecord14Service{repo: repo}
}

// Create validates and persists a new PatientRecord14.
func (s *PatientRecord14Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.PatientRecord14, error) {
	now := time.Now().UTC()
	rec := &models.PatientRecord14{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a PatientRecord14 by id.
func (s *PatientRecord14Service) Get(ctx context.Context, id string) (*models.PatientRecord14, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped PatientRecord14 pages.
func (s *PatientRecord14Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.PatientRecord14, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing PatientRecord14.
func (s *PatientRecord14Service) Update(ctx context.Context, actor string, rec *models.PatientRecord14) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.UpdatedBy = actor
	rec.UpdatedAt = time.Now().UTC()
	// repository Update applies optimistic locking and increments version
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a PatientRecord14.
func (s *PatientRecord14Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds PatientRecord14 by external code prefix.
func (s *PatientRecord14Service) Search(ctx context.Context, facilityID, prefix string) ([]models.PatientRecord14, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *PatientRecord14Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}
